specs/cli/plugin-add.md

```diff
@@ -0,0 +1,163 @@
+
+Developers can generate `openai` and `apimanifest` type of plugins. By generating `openai` or `apimanifest`, two outputs will be generated: a\) the plugin type you have chosen that will be named `{plugin-name}-{type}.json` and b\) a sliced OpenAPI description with only the endpoints that matches `--include-path` and `--exclude-path`, if provided.
+> [!NOTE] 
+> In one's solution, there might be two different [API Manifests][def]. The `apimanifest.json` in the root folder represents a single artifact surface of all APIs and it will always be generated. The second one, specific to each plugin, will be named `{plugin-name}-apimanifest.json` and saved in the chosen output directory when `apimanifest` value is used as the plugin type.
+
+Once the `workspace.json` file is generated and the OpenAPI description file is saved locally, the generation will be executed and the plugin and the sliced OpenAPI description will become available.
+
+## Parameters
+
+| Parameters | Required | Example | Description | Telemetry |
+| -- | -- | -- | -- | -- |
+| `--plugin-name \| --mn` | Yes | GitHub | Name of the plugin. Unique within the parent API. Defaults to `Plugin` | No |
+| `--openapi \| -d` | Yes | https://raw.githubusercontent.com/github/rest-api-description/main/descriptions/api.github.com/api.github.com.json | The location of the OpenAPI description in JSON or YAML format to use to generate the plugin. Accepts a URL or a local directory. | No |
+| `--search-key \| --sk` | No | apisguru::github.com:api.github.com | The search key used to locate the OpenAPI description. | Yes, without its value |
+| `--include-path \| -i` | No | /repos/{owner}/{repo} | A glob pattern to include paths from generation. Accepts multiple values. Defaults to no value which includes everything. | Yes, without its value |
+| `--exclude-path \| -e` | No | /advisories | A glob pattern to exclude paths from generation. Accepts multiple values. Defaults to no value which excludes nothing. | Yes, without its value |
+| `--type \| -t` | Yes | openai | The target type of plugin for the generated output files. Accepts multiple values. Possible values are `openai` and `apimanifest`. Defaults to `apimanifest`| Yes |
+| `--overlayDirectory \| --od` | No | ./overlays/plugins/{plugin-name}/overlay.yaml | The location of the overlay file in JSON or YAML format to be used to generate the plugin. [Overlay](https://github.com/OAI/Overlay-Specification/blob/main/versions/1.0.0.md) defines a way of creating documents that contain additional information to be merged with an OpenAPI description. Defaults to no value which uses the OpenAPI description as it is. | Yes, without its value |
```

Comment:
I'm not sure we need an additional argument for that. I haven't read the overlays spec in a while, but as far as I can remember the overlay is supposed to point to the lower overlay or the description.

Reply:
So far, the extends field in the Overlay document is not required. [Extends](https://github.com/OAI/Overlay-Specification/blob/main/versions/1.0.0.md#fixed-fields) is the field that includes the URL to the target document the overlay should be applied.
We have 2 options:
- Keep the overlay-directory as required in the plugin generation.
- Review overlay spec and make extends a required field.

cc @sebastienlevert @darrelmiller

Reply:
I don't think we should use `extends`.  I think we should follow the pattern TypeSpec uses where the "overlay" gets pulled in via the commandline in a second parameter.  This is how speakeasy has done it https://www.speakeasyapi.dev/docs/speakeasy-cli/overlay/apply

Reply:
do you expect multiple overlays to be apply at any given time? if so I believe that order will matter.
I'm not sure what'd be the best experience for people with that
```shell
kiota ... -d https://foo/bar.yaml --overlay ./ov1.json --overlay ./ov2.json
```

Doesn't make it obvious order matters

We could go the CSV approach, or even use a "structured" input like we do with structured mime types.

Thoughts?

Reply:
Could this be a simple array of overlay, and if we specify only one, it would be an array with a single item? Can that be modeled using the CLI?

Reply:
I'm not sure what you suggest would be the way to specify multiple entries, and order them?

Reply:
The same way we do it with structuredMediaTypes, right? In the workspace file, something like this : 

```
overlays: [
  "overlay1.yaml",
  "overlay2.yaml"
]

Reply:
no, the structured media types is --structured-media-types "application/json,q=1;application/yml,q=0.9;"....

Reply:
I would say we are expecting for now, one overlay to be applied. I think we can go with an unique value and if we get community feedback they would like to apply more than one, than we might consider doing the change.

Reply:
any change to the CLI is going to be breaking. I get that for now those commands are behind a feature flag, but I'd like to avoid the easy trap of:
1. designing with short term goals in mind.
2. removing the feature flag.
3. getting broad adoption and being stuck in a corner when the feedback comes.


Since the overlay support is not going to come in OAI.net before a while, what do you think about simply removing the parameter for now?

Reply:
I think this will help us move forward. Let's start without Overlay and we'll add this back in when we have more context on the use of overlays.

Reply:
@maisarissi to update the spec when you have a couple of minutes